apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,373 @@
+      if (!this._bluetooth) {
+        // roll back the setting value to notify the UIs
+        // that Bluetooth interface is not available
+        if (value) {
+          SettingsListener.getSettingsLock().set({
+            'bluetooth.enabled': false
+          });
+        }
+        return;
+      }
+    }.bind(this));
+
+    this._attributeHandler = this._bluetoothAttrChangeHandler.bind(this);
+
+    // watch attributechanged event
+    this._bluetooth.addEventListener('attributechanged',
+      this._attributeHandler);
+
+    // clear defaultAdapter and listener once adapter is removed
+    this._bluetooth.addEventListener('adapterremoved', () => {
+      this._adapter = null;
+      this._bluetooth.removeEventListener('attributechanged',
+        this._attributeHandler);
+    });
+
+    // if bluetooth is enabled in booting time, try to get adapter now
+    this.debug('init bluetooth adapter');
+    this._adapter = this._bluetooth.defaultAdapter;
+    this._adapter.addEventListener('attributechanged',
+      this._bluetoothAdapterAttrChangeHandler.bind(this));
+
+    /* In file transfering case:
+     * since System Message can't be listened in two js files within a app,
+     * so we listen here but dispatch events to bluetooth_transfer.js
+     * while getting bluetooth file transfer start/complete system messages
+     */
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-start',
+      this._relayMessageEvent.bind(this, 'bluetooth-opp-transfer-start')
+    );
+
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-complete',
+      this._relayMessageEvent.bind(this, 'bluetooth-opp-transfer-complete')
+    );
+
+    // decouple bluetooth enable/disable function from other system part
+    window.addEventListener('request-enable-bluetooth',
+      this._enableHandler.bind(this));
+    window.addEventListener('request-disable-bluetooth',
+      this._disableHandler.bind(this));
+
+    Service.registerState('isEnabled', this);
+    // LazyLoader.load(['js/bluetooth_icon.js',
+    //                  'js/bluetooth_transfer_icon.js',
+    //                  'js/bluetooth_headphone_icon.js'], function() {
+    //   this.icon = new BluetoothIcon(this);
+    //   this.icon.start();
+    //   this.transferIcon = new BluetoothTransferIcon(this);
+    //   this.transferIcon.start();
+    //   this.headphoneIcon = new BluetoothHeadphoneIcon(this);
+    //   this.headphoneIcon.start();
+    // }.bind(this)).catch(function(err) {
+    //   console.error(err);
+    // });
+  },
+
+  /**
+   * update settings value and enable bluetooth hardware
+   *
+   * @private
+   */
+  _enableHandler: function bt__enableHandler() {
+    this.debug('enable bluetooth');
+    this.getAdapter().then((adapter) => {
+      adapter.enable().then(() => { //resolve
+        this._updateProfileStat(adapter);
+        navigator.mozSettings.createLock().set({
+          'bluetooth.enabled': true
+        });
```

Comment:
I would like to suggest you update mozSettings key `bluetooth.enabled` in the adapter `attributechanged` function. In my experience, the event will come more quickly than the `adapter.enable()` promise resolved here.